common/src/main/java/com/alibaba/nacos/common/lifecycle/ResourceLifeCycleManager.java

```diff
@@ -0,0 +1,141 @@
+
+    private static final ResourceLifeCycleManager INSTANCE = new ResourceLifeCycleManager();
+
+    private static final AtomicBoolean CLOSED = new AtomicBoolean(false);
+
+    static {
+        INSTANCE.init();
+        ShutdownUtils.addShutdownHook(new Thread(new Runnable() {
+            @Override
+            public void run() {
+                LOGGER.warn("[LifeCycleManager] Start destroying Every Instance");
+                shutdown();
+                LOGGER.warn("[LifeCycleManager] Destruction of the end");
+            }
+        }));
+    }
+
+    private void init() {
+        this.lifeCycleResources = new CopyOnWriteArrayList<AbstractLifeCycle>();
+    }
+
+    public static ResourceLifeCycleManager getInstance() {
+        return INSTANCE;
+    }
+
+    /**
+     * Destroy all of the life cycle resources which are managed by ResourceLifeCycleManager.
+     *
+     */
+    public static void shutdown() {
+        if (!CLOSED.compareAndSet(false, true)) {
+            return;
+        }
+
+        List<AbstractLifeCycle> instances = INSTANCE.lifeCycleResources;
+        for (AbstractLifeCycle instance : instances) {
+            INSTANCE.destroy(instance);
+        }
+    }
+
+    /**
+     * Destroy the life cycle resource instance.
+     *
+     * @param instance the life cycle resource instance which is need to be destroyed.
+     */
+    public void destroy(AbstractLifeCycle instance) {
+        final Object monitor = lockers.get(instance);
+        if (monitor == null) {
+            return;
+        }
+        synchronized (monitor) {
+            try {
+                // the life cycle resources which managed are do stop method.
+                if (LOGGER.isDebugEnabled()) {
+                    LOGGER.debug("Life cycle resources do stop");
+                }
+                instance.stop();
+                INSTANCE.lifeCycleResources.remove(instance);
+                INSTANCE.lockers.remove(instance);
+            } catch (Exception e) {
+                LOGGER.error("An exception occurred during resource do stop : {}", e);
+            }
+        }
+    }
+
+
+    /**
+     * Cancel the specified lifecycle resource instance management.
+     *
+     * @param instance the management life cycle resource instances;
+     *
+     */
+    public void deregister(AbstractLifeCycle instance) {
+        if (this.lifeCycleResources.contains(instance)) {
+            final Object monitor = lockers.get(instance);
+            synchronized (monitor) {
+                this.lifeCycleResources.remove(instance);
+                this.lockers.remove(instance);
+            }
+        }
+    }
+
+    /**
+     * Register the life cycle resource instances into the lifeCycleResources lists.
+     *
+     * @param instance the management life cycle resource instances.
+     */
+    public void register(AbstractLifeCycle instance) {
+        if (!lifeCycleResources.contains(instance)) {
+            synchronized(this) {
```

Comment:
New Object() may be executed multiple times

Reply:
```java
if (!lifeCycleResources.contains(instance)) {
            synchronized(this) {
                          if (!lockers.containsKey(instance)) {
                                      lockers.put(instance, new Objects());
                          }
             }
}
```

Reply:
Okay I will resolve this issue in the codes.